core/libp2p/security/plaintext/plaintext.cpp

```diff
@@ -5,69 +5,134 @@
-                      session](crypto::PublicKey remoteKey) mutable {
-      // when we received remote public key, send them our public key
-      session->sendKey(
-          this->idmgr_->getKeyPair().publicKey,
-          [this, cb{std::move(cb)}, inbound{std::move(inbound)}, session,
-           remoteKey{std::move(remoteKey)}]() mutable {
-            auto plaintext = std::make_shared<connection::PlaintextConnection>(
-                inbound, this->idmgr_->getKeyPair().publicKey, remoteKey);
-
-            return cb(std::move(plaintext));
-          });
-    });
+    sendExchangeMsg(inbound, cb);
+    receiveExchangeMsg(inbound, cb);
   }
 
   void Plaintext::secureOutbound(
       std::shared_ptr<connection::RawConnection> outbound,
-      const peer::PeerId &p, SecConnCallbackFunc cb) {
-    auto session =
-        std::make_shared<PlaintextSession>(this->marshaller_, outbound, cb);
-
-    // immediately send our public key
-    session->sendKey(
-        this->idmgr_->getKeyPair().publicKey,
-        [this, session, outbound{std::move(outbound)}, p,
-         cb{std::move(cb)}]() mutable {
-          // then wait for key from other side
-          session->recvKey([this, p, outbound{std::move(outbound)},
-                            cb{std::move(cb)}](
-                               crypto::PublicKey remotePubkey) mutable {
-            auto remoteIdCalculated = peer::PeerId::fromPublicKey(remotePubkey);
-            if (remoteIdCalculated != p) {
-              return cb(make_error_code(SecurityError::AUTHENTICATION_ERROR));
-            }
+      const peer::PeerId &p,
+      SecConnCallbackFunc cb) {
+    sendExchangeMsg(outbound, cb);
+    receiveExchangeMsg(outbound, cb);
+  }
 
-            auto plaintext = std::make_shared<connection::PlaintextConnection>(
-                outbound, this->idmgr_->getKeyPair().publicKey, remotePubkey);
+  void Plaintext::sendExchangeMsg(
+      std::shared_ptr<connection::RawConnection> conn,
+      SecConnCallbackFunc cb) const {
+    PLAINTEXT_OUTCOME_TRY(out_msg,
+                          marshaller_->marshal(plaintext::ExchangeMessage{
+                              .pubkey = idmgr_->getKeyPair().publicKey,
+                              .peer_id = idmgr_->getId()}),
+                          conn,
+                          cb);
+    conn->write(out_msg.value(), out_msg.value().size(), [cb{std::move(cb)}, conn](auto &&res) {
+      if (res.has_error()) {
+        conn->close();
+        cb(Error::EXCHANGE_SEND_ERROR);
+      }
+    });
+  }
 
-            return cb(std::move(plaintext));
-          });
-        });
+  void Plaintext::receiveExchangeMsg(
+      const std::shared_ptr<connection::RawConnection> &conn,
+      SecConnCallbackFunc cb) const {
+    constexpr size_t kMaxMsgSize = 10000;
+    auto read_bytes = std::make_shared<std::vector<uint8_t>>(kMaxMsgSize);
+    conn->readSome(*read_bytes,
+                   kMaxMsgSize,
+                   std::bind(&Plaintext::readCallback,
```

Comment:
IMO, better use lambda, since it would be more clear.

Reply:
Ok. If you think so.